### PR TITLE
Remove pp/cpp name length limit

### LINKED ISSUE #4711: [Umbrella] Use permanent-id to replace namespace/name labels in the resource
**What would you like to be added**:

Please ref to https://github.com/karmada-io/karmada/issues/4683#issuecomment-1996604867

**Why is this needed**:

Continue the work of #4000 and break down the tasks to be done.

- [x] Unify and add `permanent-id` labels to `PropagationPolicy/ClusterPropagationPolicy`, `ResourceBinding/ClusterResourceBinding`, and `Work` resources in `karmada-webhook`.
  - [x] Add `permanent-id` for pp, cpp, rb, crb, work in the karmada-webhook @jwcesign @XiShanYongYe-Chang #4474 
  - [x] Remove the add `permanent-id` logic in the karmada-controller-manager (we can do it in the next release) @whitewindmills 
    - [x] For Work resource #4928 
    - [x] For RB/CRB resource #4950 
    - [x] For PP/CPP resource #4949 
- [x] Replace the use of `namespace/name` labels in the code with `permanent-id` labels, including: @whitewindmills @XiShanYongYe-Chang 
  - [x] `propagationpolicy.karmada.io/namespace`, `propagationpolicy.karmada.io/name` -> `propagationpolicy.karmada.io/permanent-id` #4743 
  - [x] `clusterpropagationpolicy.karmada.io/name` -> `clusterpropagationpolicy.karmada.io/permanent-id` #4743 
  - [x] `resourcebinding.karmada.io/key` -> `resourcebinding.karmada.io/permanent-id` #4811 
  - [x] `clusterresourcebinding.karmada.io/key` -> `clusterresourcebinding.karmada.io/permanent-id` #4811 
  - [x] `work.karmada.io/namespace`, `work.karmada.io/name` -> `work.karmada.io/permanent-id` #4813 #4765 #4865 
- [x] Clean up the `uid` labels on the resource: @liangyuanpeng #4718 
  - [x] `propagationpolicy.karmada.io/uid`
  - [x] `clusterpropagationpolicy.karmada.io/uid`
  - [x] `resourcebinding.karmada.io/uid`
  - [x] `clusterresourcebinding.karmada.io/uid`
  - [x] `work.karmada.io/uid`
- [x] Deprecated `namespace/name` labels on resources: @whitewindmills
  - [x] Deprecated `propagationpolicy.karmada.io/namespace` #4743
  - [x] Deprecated `propagationpolicy.karmada.io/name` #4743 
  - [x] Deprecated `clusterpropagationpolicy.karmada.io/name` #4743 
  - [x] `resourcebinding.karmada.io/key` #4811 
  - [x] `clusterresourcebinding.karmada.io/key`  #4811 
  - [x] `work.karmada.io/namespace` #4813 
  - [x] `work.karmada.io/name` #4813 
- [x] Add validation for permanent-id @whitewindmills #4964
- [x]  Remove those labels: `propagationpolicy.karmada.io/namespace`, `propagationpolicy.karmada.io/name` in the v1.11. @XiShanYongYe-Chang #5020 
- [x] The names of PP and CPP are not allowed to exceed 63 characters. This is the verification logic of karmada-webhook: 
https://github.com/karmada-io/karmada/blob/9822d099d1b717902003b2c7a11448d2cea6ac05/pkg/webhook/propagationpolicy/mutating.go#L74-L76
@XiShanYongYe-Chang #5029
> Note: the reason for adding this verification is that the names of PP and CPP will appear in In the resource template and the label value of RB, this check can be removed after we solve the issue.
- [x] Describe the impact of the current task on users and upgrade precautions in the upgrade document. @zhzhuang-zju https://github.com/karmada-io/website/pull/622

### PR DESCRIPTION
**Pull request by XiShanYongYe-Chang**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

In the previous release, the names of PP and CPP are not allowed to exceed 63 characters. Now, we remove this limit.

**Which issue(s) this PR fixes**:
Fix #4711

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
karmada-webhook: Remove the limit of 63 name lengths with PropagationPolicy/ClusterPropagationPolicy resource.
```

### CODE REVIEW

**Comment by XiShanYongYe-Chang** [comment]
Test report:
image

**Comment by zhzhuang-zju** [comment]
If this is the last task, changing `part of` to `fix` would be better

File: pkg/webhook/clusterpropagationpolicy/mutating.go [hunk 1]
```diff
@@ -66,10 +64,6 @@ func (a *MutatingAdmission) Handle(_ context.Context, req admission.Request) adm
 	helper.AddTolerations(&policy.Spec.Placement, helper.NewNotReadyToleration(a.DefaultNotReadyTolerationSeconds),
 		helper.NewUnreachableToleration(a.DefaultUnreachableTolerationSeconds))
 
-	if len(policy.Name) > validation.LabelValueMaxLength {
```
Comment (chaunceyjiang) [inline pkg/webhook/clusterpropagationpolicy/mutating.go:69]: Just one question: Do we have e2e to test this case now?

The length of pp's name exceeds 63.

**Review by whitewindmills — COMMENTED** [review]
/lgtm

File: pkg/webhook/clusterpropagationpolicy/mutating.go (hunk 1 above)
Comment (XiShanYongYe-Chang) [reply pkg/webhook/clusterpropagationpolicy/mutating.go:69]: name is so long, it may not be a common test case.

Comment (chaunceyjiang) [reply pkg/webhook/clusterpropagationpolicy/mutating.go:69]: I suggest adding a case for testing.

Comment (XiShanYongYe-Chang) [reply pkg/webhook/clusterpropagationpolicy/mutating.go:69]: Ok~

Comment (XiShanYongYe-Chang) [reply pkg/webhook/clusterpropagationpolicy/mutating.go:70]: Update, please help a review again.

File: test/e2e/clusterpropagationpolicy_test.go [hunk 2]
```diff
@@ -547,8 +599,12 @@ var _ = ginkgo.Describe("[AdvancedClusterPropagation] propagation testing", func
 })
 
 // ImplicitPriority more than one PP matches the object, we should choose the most suitable one.
+<<<<<<< HEAD
 // Set it to run sequentially to avoid affecting other test cases.
 var _ = framework.SerialDescribe("[ImplicitPriority] propagation testing", func() {
+=======
+var _ = ginkgo.Describe("[ImplicitPriority] ClusterPropagationPolicy testing", func() {
+>>>>>>> 97207eefc (remove pp/cpp name length limit)
```
Comment (liangyuanpeng) [inline test/e2e/clusterpropagationpolicy_test.go:607]: I guess you forgot to handle this.

Comment (XiShanYongYe-Chang) [reply test/e2e/clusterpropagationpolicy_test.go]: Thanks for your reminder!

File: test/e2e/propagationpolicy_test.go [hunk 3]
```diff
@@ -513,9 +491,67 @@ var _ = ginkgo.Describe("[BasicPropagation] propagation testing", func() {
...
+		})
+	})
+
+	ginkgo.Context("Propagate Deployment with long pp name (exceed 63 character)", func() {
+		var deploymentNamespace, deploymentName string
+		var deployment *appsv1.Deployment
+
+		ginkgo.BeforeEach(func() {
+			policyNamespace = testNamespace
+			policyName = deploymentNamePrefix + "-longname-longname-longname-longname-longname-longname-" + rand.String(RandomStrLength)
```
Comment (liangyuanpeng) [inline test/e2e/propagationpolicy_test.go:511]: I think need to  verify `policyName` length > 63

Comment (XiShanYongYe-Chang) [reply test/e2e/propagationpolicy_test.go:511]: Yes, the length of the policy name exceeds 63 characters.

**Comment by XiShanYongYe-Chang** [comment]
Hi @chaunceyjiang can you help take a review?
/assign @chaunceyjiang

**Comment by XiShanYongYe-Chang** [comment]
cc @RainbowMango 
Can we bring this into this version?

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Pull request merged**